### DISCUSSION TITLE
Adding .required does not make returned type have required field

### DISCUSSION THREAD

**Discussion by robogeek** [colinhacks/zod#3071] [Q&A] [answered]
I have a type with a required field, for which I've generated a Zod schema.  The generated schema is coming from `openapi-to-zod`, but the small example below replicates the relevant code and demonstrates the problem.  In the documentation, the section on Objects leads off with sample code in which it is said that by default fields of `z.object` are required, and that the example is equivalent to a TypeScript definition where the fields are required.

As shown below, that's not the case today.

```js
import { z } from 'zod';

type E = {
    programID: string;
    eName?: string;
};

// Establish that programID is required
const badE: E = {};
// Property 'programID' is missing in type '{}' but required in type 'E'.
const goodE: E = { programID: '99' }
// No error

const parseE = z.object({
    programID: z
                .string()
                .regex(new RegExp("^[a-zA-Z0-9_-]*$"))
                .min(1)
                .max(128)
                .describe("URL safe VTN assigned object ID."),
    eName: z.string().default('ZILCH').optional()
}); // .required(); // { programID: true });

// Generate Zod's view of the type described by the schema
type tE = z.infer<typeof parseE>;

// This is the type shown when hovering the mouse over `tE`
// Notice that programID is shown as optional.
// No addition of .required() in the parseE definition changes this
//
// type tE = {
//     programID?: string;
//     eName?: string;
// }

// tE should require programID, therefore this should give
// the same error as above
const badE2: tE = {};
// No error

const e: E = parseE.parse({
    // Commenting this out results in a parse error
    // saying that programID is required 
    programID: 'Program-ID-99'
});

console.log(e);
```

The field `E.programID` is required.  In `parseE` I'm trying to work out how to make `programID` be required as well.

Compiling the above gives the following error:

```
src/index.ts:30:7 - error TS2322: Type '{ programID?: string; eName?: string; }' is not assignable to type 'E'.
  Property 'programID' is optional in type '{ programID?: string; eName?: string; }' but required in type 'E'.

30 const e: E = parseE.parse({
```

The error message says the Zod schema returned by `parseE.parse` has `programID` as optional.  TypeScript complains that it is not assignable because `programID` is required.

No matter the setup of the `.required()` call, the error message does not change.

In Visual Studio Code, and hovering the mouse over things - the `parseE.parse` function shows `programID` as optional.

As noted in the comments, using `z.infer` to retrieve the type Zod thinks is correct shows `programID` to be optional.  No option for `.required()` changes this.

But, if the object passed to `parseE.parse` does not have `programID`, then when the program is run an error is thrown with this issue:

```js
{
      code: 'invalid_type',
      expected: 'string',
      received: 'undefined',
      path: [ 'programID' ],
      message: 'Required'
    }
```

So.. the parser knows that `programID` is required, while TYpeScript compiler is still giving the error saying `programID` is optional.

**Comment by robogeek** [comment, marked as answer]
Hurm.... I noticed my `tsconfig.json` did not have `"strict": true`, added that, and now the code is working.

**Comment by JacobWeisenburger** [comment]
It looks like your question is answered, can I close this issue?

**Comment by robogeek** [comment]
Yes
